File: paleo_workbench/ui/workstation/app_bar.py
from __future__ import annotations

from PySide6.QtCore import Qt, Signal
from PySide6.QtGui import QActionGroup
from PySide6.QtWidgets import (
    QFrame,
    QHBoxLayout,
    QLabel,
    QLineEdit,
    QMenu,
    QSizePolicy,
    QToolButton,
)

from paleo_workbench import tokens
from paleo_workbench.ui.workstation.common import workstation_icon


class WorkstationAppBar(QFrame):
    """Global-only actions and project context for the workstation shell."""

    new_project_requested = Signal()
    open_project_requested = Signal()
    open_sample_requested = Signal()
    save_project_requested = Signal()
    properties_requested = Signal()
    command_submitted = Signal(str)
    agent_requested = Signal()
    task_center_requested = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("WorkstationAppBar")
        self.setFixedHeight(46)

        layout = QHBoxLayout(self)
        layout.setContentsMargins(10, 0, 10, 0)
        layout.setSpacing(8)

        brand_icon = QLabel(self)
        brand_icon.setObjectName("WorkstationBrandIcon")
        brand_icon.setPixmap(
            workstation_icon("seismic.svg", tokens.PRIMARY).pixmap(20, 20)
        )
        layout.addWidget(brand_icon)

        self.brand_label = QLabel("Paleo Workbench", self)
        self.brand_label.setObjectName("WorkstationBrand")
        layout.addWidget(self.brand_label)

        self.project_button = QToolButton(self)
        self.project_button.setObjectName("WorkstationProjectButton")
        self.project_button.setToolButtonStyle(Qt.ToolButtonStyle.ToolButtonTextBesideIcon)
        self.project_button.setPopupMode(QToolButton.ToolButtonPopupMode.InstantPopup)
        self.project_button.setIcon(workstation_icon("folder-open.svg"))
        self._project_menu = QMenu(self.project_button)
        self._project_menu.addAction("新建工程", self.new_project_requested.emit)
        self._project_menu.addAction("打开工程", self.open_project_requested.emit)
        self._project_menu.addAction("打开示例", self.open_sample_requested.emit)
        self._project_menu.addSeparator()
        self._project_menu.addAction("保存工程", self.save_project_requested.emit)
        self._project_menu.addAction("工程属性", self.properties_requested.emit)
        self.project_button.setMenu(self._project_menu)
        layout.addWidget(self.project_button)

        self.back_button = QToolButton(self)
        self.back_button.setObjectName("WorkstationChromeButton")
        self.back_button.setIcon(workstation_icon("arrow-left.svg"))
        self.back_button.setToolTip("后退")
        self.back_button.setEnabled(False)
        layout.addWidget(self.back_button)

        self.forward_button = QToolButton(self)
        self.forward_button.setObjectName("WorkstationChromeButton")
        self.forward_button.setIcon(workstation_icon("arrow-right.svg"))
        self.forward_button.setToolTip("前进")
        self.forward_button.setEnabled(False)
        layout.addWidget(self.forward_button)

        layout.addStretch(1)

        self.command_input = QLineEdit(self)
        self.command_input.setObjectName("WorkstationCommandInput")
        self.command_input.setPlaceholderText("搜索命令、数据或输入 Agent 指令 (Ctrl+K)")
        self.command_input.setClearButtonEnabled(True)
        self.command_input.setMinimumWidth(300)
        self.command_input.setMaximumWidth(580)
        self.command_input.setSizePolicy(
            QSizePolicy.Policy.Expanding, QSizePolicy.Policy.Fixed
        )
        self.command_input.returnPressed.connect(self._submit_command)
        layout.addWidget(self.command_input, 2)

        layout.addStretch(1)

        # B1/B16：视图菜单是主题与密度的生产入口（此前 set_theme 只有测试
        # 调用，暗色/高对比在生产不可达；Ribbon 密度按钮随 Ribbon 0 高不可见）。
        self.view_button = QToolButton(self)
        self.view_button.setObjectName("WorkstationChromeButton")
        self.view_button.setIcon(workstation_icon("rb-density-comfortable.svg"))
        self.view_button.setText("视图")
        self.view_button.setToolButtonStyle(Qt.ToolButtonStyle.ToolButtonTextBesideIcon)
        self.view_button.setPopupMode(QToolButton.ToolButtonPopupMode.InstantPopup)
        self.view_button.setToolTip("主题与界面密度")
        self._view_menu = QMenu(self.view_button)
        self._theme_group = QActionGroup(self._view_menu)
        self._density_group = QActionGroup(self._view_menu)
        from paleo_workbench.ui.theme import DensityMode, ThemeMode, theme_manager

        for mode, label in (
            (ThemeMode.LIGHT, "浅色主题"),
            (ThemeMode.DARK, "深色主题"),
            (ThemeMode.HIGH_CONTRAST, "高对比主题"),
        ):
            action = self._view_menu.addAction(label)
            action.setCheckable(True)
            action.setChecked(theme_manager.current_theme is mode)
            action.setActionGroup(self._theme_group)
            action.triggered.connect(
                lambda _checked=False, m=mode: theme_manager.set_theme(m)
            )
        self._view_menu.addSeparator()
        for mode, label in (
            (DensityMode.COMPACT, "紧凑密度"),
            (DensityMode.COMFORTABLE, "舒适密度"),
        ):
            action = self._view_menu.addAction(label)
            action.setCheckable(True)
            action.setChecked(theme_manager.density is mode)
            action.setActionGroup(self._density_group)
            action.triggered.connect(
                lambda _checked=False, m=mode: theme_manager.set_density(m)
            )
        # 主题/密度变化时刷新勾选态（跨入口切换保持一致）。
        def _sync_view_checks(*_args) -> None:
            for action, mode in zip(self._theme_group.actions(), list(ThemeMode)):
                action.setChecked(theme_manager.current_theme is mode)
            for action, mode in zip(
                self._density_group.actions(), list(DensityMode)
            ):
                action.setChecked(theme_manager.density is mode)

        theme_manager.theme_changed.connect(_sync_view_checks)
        self.view_button.setMenu(self._view_menu)
        layout.addWidget(self.view_button)

        self.task_button = QToolButton(self)
        self.task_button.setObjectName("WorkstationTaskButton")
        self.task_button.setIcon(workstation_icon("rb-run.svg"))
        self.task_button.setText("任务 0")
        self.task_button.setToolButtonStyle(Qt.ToolButtonStyle.ToolButtonTextBesideIcon)
        self.task_button.setToolTip("打开任务中心")
        self.task_button.clicked.connect(self.task_center_requested.emit)
        layout.addWidget(self.task_button)

        self.agent_button = QToolButton(self)
        self.agent_button.setObjectName("WorkstationAgentButton")
        self.agent_button.setIcon(workstation_icon("visualization.svg"))
        self.agent_button.setText("Agent")
        self.agent_button.setToolButtonStyle(Qt.ToolButtonStyle.ToolButtonTextBesideIcon)
        self.agent_button.setToolTip("打开上下文感知 Agent 工作区")
        self.agent_button.clicked.connect(self.agent_requested.emit)
        layout.addWidget(self.agent_button)

        self.set_project("未命名工程", "")

    def set_project(self, name: str, region: str = "") -> None:
        project = str(name or "未命名工程")
        area = str(region or "").strip()
        if area.casefold() == project.casefold():
            area = ""
        self._project_region = area
        self.project_button.setText(f"{project}  /  {area}" if area else project)
        self.project_button.setToolTip("切换工程或打开工程操作")

    def set_project_name(self, name: str) -> None:
        self.set_project(name, getattr(self, "_project_region", ""))

    def set_task_count(self, active: int) -> None:
        count = max(0, int(active))
        self.task_button.setText(f"任务 {count}")
        self.task_button.setProperty("activeTasks", count > 0)
        self.task_button.style().unpolish(self.task_button)
        self.task_button.style().polish(self.task_button)

    def focus_command(self) -> None:
        self.command_input.setFocus(Qt.FocusReason.ShortcutFocusReason)
        self.command_input.selectAll()

    def _submit_command(self) -> None:
        text = self.command_input.text().strip()
        if not text:
            return
        self.command_input.clear()
        self.command_submitted.emit(text)

File: libs/ui_workstation/src/app_bar.cpp
#include "pwb/ui_workstation/app_bar.hpp"

#include <QActionGroup>
#include <QHBoxLayout>
#include <QLabel>
#include <QSizePolicy>

#include <pwb/ui_shell/dock_registry.hpp>
#include <pwb/ui_shell/layout_presets.hpp>
#include <pwb/ui_widgets/icon_factory.hpp>

namespace pwb::ui_workstation {

WorkstationAppBar::WorkstationAppBar(QWidget* parent) : QFrame(parent) {
    setObjectName("WorkstationAppBar");
    setFixedHeight(46);  // comfortable density default; host may rebind

    auto* layout = new QHBoxLayout(this);
    layout->setContentsMargins(8, 0, 8, 0);
    layout->setSpacing(8);

    auto* brand_icon = new QLabel(this);
    brand_icon->setObjectName("WorkstationBrandIcon");
    brand_icon->setPixmap(
        ui_widgets::workstation_icon("seismic.svg").pixmap(20, 20));
    layout->addWidget(brand_icon);

    auto* brand_label = new QLabel("Paleo Workbench", this);
    brand_label->setObjectName("WorkstationBrand");
    layout->addWidget(brand_label);

    project_button_ = new QToolButton(this);
    project_button_->setObjectName("WorkstationProjectButton");
    project_button_->setToolButtonStyle(
        Qt::ToolButtonStyle::ToolButtonTextBesideIcon);
    project_button_->setPopupMode(
        QToolButton::ToolButtonPopupMode::InstantPopup);
    project_button_->setIcon(
        ui_widgets::workstation_icon("folder-open.svg"));
    auto* project_menu = new QMenu(project_button_);
    project_menu->addAction("新建工程", this,
                            [this] { emit new_project_requested(); });
    project_menu->addAction("打开工程", this,
                            [this] { emit open_project_requested(); });
    project_menu->addAction("打开示例", this,
                            [this] { emit open_sample_requested(); });
    project_menu->addSeparator();
    project_menu->addAction("保存工程", this,
                            [this] { emit save_project_requested(); });
    project_menu->addAction("工程属性", this,
                            [this] { emit properties_requested(); });
    project_button_->setMenu(project_menu);
    layout->addWidget(project_button_);

    // 工作区预设（B2）：「自定义」+ 注册表预设，选中有 id 的项才发射。
    workspace_combo_ = new QComboBox(this);
    workspace_combo_->setObjectName("WorkstationWorkspaceCombo");
    workspace_combo_->setToolTip("切换工作区布局预设");
    set_workspace_presets(ui_shell::preset_labels());
    connect(workspace_combo_, &QComboBox::currentIndexChanged, this,
            [this](int index) {
                if (index < 0 ||
                    index >= static_cast<int>(workspace_ids_.size())) {
                    return;
                }
                const std::string& id = workspace_ids_[index];
                if (!id.empty()) {
                    emit workspace_preset_requested(
                        QString::fromStdString(id));
                }
            });
    layout->addWidget(workspace_combo_);

    layout->addStretch(1);

    command_input_ = new QLineEdit(this);
    command_input_->setObjectName("WorkstationCommandInput");
    command_input_->setPlaceholderText(
        "搜索命令、数据或输入 Agent 指令 (Ctrl+K)");
    command_input_->setClearButtonEnabled(true);
    command_input_->setMinimumWidth(180);
    command_input_->setMaximumWidth(580);
    command_input_->setSizePolicy(QSizePolicy::Policy::Expanding,
                                  QSizePolicy::Policy::Fixed);
    connect(command_input_, &QLineEdit::returnPressed, this,
            &WorkstationAppBar::submit_command);
    layout->addWidget(command_input_, 2);

    layout->addStretch(1);

    // 视图菜单（B1/B16）：主题与密度的生产入口。
    auto* view_button = new QToolButton(this);
    view_button->setObjectName("WorkstationChromeButton");
    view_button->setIcon(
        ui_widgets::workstation_icon("rb-density-comfortable.svg"));
    view_button->setText("视图");
    view_button->setToolButtonStyle(
        Qt::ToolButtonStyle::ToolButtonTextBesideIcon);
    view_button->setPopupMode(
        QToolButton::ToolButtonPopupMode::InstantPopup);
    view_button->setToolTip("主题与界面密度");
    view_menu_ = new QMenu(view_button);
    view_button->setMenu(view_menu_);
    layout->addWidget(view_button);
    // Default rows (host may rebind via set_*_choices); separators act
    // as insertion anchors so rebinds keep the menu order.
    set_theme_choices({{"light", "浅色主题"},
                       {"dark", "深色主题"},
                       {"high_contrast", "高对比主题"}},
                      "light");
    theme_separator_ = view_menu_->addSeparator();
    set_density_choices(
        {{"compact", "紧凑密度"}, {"comfortable", "舒适密度"}},
        "comfortable");
    about_separator_ = view_menu_->addSeparator();
    view_menu_->addAction("关于", this,
                          [this] { emit about_requested(); });

    task_button_ = new QToolButton(this);
    task_button_->setObjectName("WorkstationTaskButton");
    task_button_->setIcon(ui_widgets::workstation_icon("rb-run.svg"));
    task_button_->setToolButtonStyle(
        Qt::ToolButtonStyle::ToolButtonTextBesideIcon);
    task_button_->setToolTip("打开任务中心");
    connect(task_button_, &QToolButton::clicked, this,
            [this] { emit task_center_requested(); });
    layout->addWidget(task_button_);
    set_task_count(0);

    agent_button_ = new QToolButton(this);
    agent_button_->setObjectName("WorkstationAgentButton");
    agent_button_->setIcon(
        ui_widgets::workstation_icon("visualization.svg"));
    agent_button_->setText("Agent");
    agent_button_->setToolButtonStyle(
        Qt::ToolButtonStyle::ToolButtonTextBesideIcon);
    agent_button_->setToolTip("打开上下文感知 Agent 工作区");
    connect(agent_button_, &QToolButton::clicked, this,
            [this] { emit agent_requested(); });
    layout->addWidget(agent_button_);

    set_project("未命名工程", "");
}

void WorkstationAppBar::set_workspace_presets(
    std::vector<std::pair<std::string, std::string>> presets) {
    workspace_combo_->blockSignals(true);
    workspace_combo_->clear();
    workspace_ids_.clear();
    workspace_ids_.push_back("");
    workspace_combo_->addItem("自定义");
    for (const auto& [id, label] : presets) {
        workspace_ids_.push_back(id);
        workspace_combo_->addItem(QString::fromStdString(label));
    }
    workspace_combo_->blockSignals(false);
}

void WorkstationAppBar::set_current_workspace(
    const std::string& preset_id) {
    int index = 0;
    for (std::size_t i = 0; i < workspace_ids_.size(); ++i) {
        if (workspace_ids_[i] == preset_id) {
            index = static_cast<int>(i);
            break;
        }
    }
    workspace_combo_->blockSignals(true);
    workspace_combo_->setCurrentIndex(index);
    workspace_combo_->blockSignals(false);
}

void WorkstationAppBar::set_project(const QString& name,
                                    const QString& region) {
    const QString project = name.isEmpty() ? "未命名工程" : name;
    QString area = region.trimmed();
    if (area.compare(project, Qt::CaseInsensitive) == 0) {
        area.clear();
    }
    project_region_ = area;
    project_button_->setText(
        area.isEmpty() ? project : project + "  /  " + area);
    project_button_->setToolTip("切换工程或打开工程操作");
}

void WorkstationAppBar::set_project_name(const QString& name) {
    set_project(name, project_region_);
}

void WorkstationAppBar::set_task_count(int active) {
    const int count = std::max(0, active);
    task_button_->setText(QString("任务 %1").arg(count));
    task_button_->setProperty("activeTasks", count > 0);
    task_button_->style()->unpolish(task_button_);
    task_button_->style()->polish(task_button_);
}

void WorkstationAppBar::set_command_input_floor(int floor_px) {
    command_input_->setMinimumWidth(floor_px);
}

void WorkstationAppBar::set_theme_choices(
    const std::vector<std::pair<std::string, std::string>>& choices,
    const std::string& current) {
    // Rebuild the theme block at the top of the view menu.
    for (auto& [action, _] : theme_actions_) {
        view_menu_->removeAction(action);
        action->deleteLater();
    }
    theme_actions_.clear();
    for (const auto& [value, label] : choices) {
        QAction* action = new QAction(QString::fromStdString(label),
                                      view_menu_);
        connect(action, &QAction::triggered, this, [this, value] {
            emit theme_requested(QString::fromStdString(value));
        });
        action->setCheckable(true);
        action->setChecked(value == current);
        // Theme block opens the menu; on the first call the separator
        // doesn't exist yet — appending is equivalent then.
        if (theme_separator_ != nullptr) {
            view_menu_->insertAction(theme_separator_, action);
        } else {
            view_menu_->addAction(action);
        }
        theme_actions_.emplace_back(action, value);
    }
}

void WorkstationAppBar::set_density_choices(
    const std::vector<std::pair<std::string, std::string>>& choices,
    const std::string& current) {
    for (auto& [action, _] : density_actions_) {
        view_menu_->removeAction(action);
        action->deleteLater();
    }
    density_actions_.clear();
    for (const auto& [value, label] : choices) {
        QAction* action = new QAction(QString::fromStdString(label),
                                      view_menu_);
        connect(action, &QAction::triggered, this, [this, value] {
            emit density_requested(QString::fromStdString(value));
        });
        action->setCheckable(true);
        action->setChecked(value == current);
        // Density block sits between the two separators; on the first
        // construction call the about separator doesn't exist yet —
        // appending puts the action in the same slot.
        if (about_separator_ != nullptr) {
            view_menu_->insertAction(about_separator_, action);
        } else {
            view_menu_->addAction(action);
        }
        density_actions_.emplace_back(action, value);
    }
}

void WorkstationAppBar::sync_view_checks(
    const std::string& current_theme, const std::string& current_density) {
    for (auto& [action, value] : theme_actions_) {
        action->setChecked(value == current_theme);
    }
    for (auto& [action, value] : density_actions_) {
        action->setChecked(value == current_density);
    }
}

void WorkstationAppBar::focus_command() {
    command_input_->setFocus(Qt::FocusReason::ShortcutFocusReason);
    command_input_->selectAll();
}

void WorkstationAppBar::submit_command() {
    const QString text = command_input_->text().trimmed();
    if (text.isEmpty()) return;
    command_input_->clear();
    emit command_submitted(text);
}

}  // namespace pwb::ui_workstation
